### PR TITLE
chore: parse and write the number of unchanged items in a sequence

### PR DESCRIPTION
**Pull request by Lou1415926**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/template/diff/write.go [hunk 1]
```diff
@@ -38,9 +40,12 @@ func (s *treeWriter) write() error {
...
 func (s *treeWriter) writeTree(node diffNode, indent int) error {
 	var formatter formatter
-	switch node.(type) {
-	// case *unchangedNode:
-	// TODO(lou1425926): handle unchanged. 
+	switch node := node.(type) {
+	case *unchangedNode:
+		content := fmt.Sprintf("(%s)", english.Plural(node.unchangedCount(), "unchanged item", "unchanged items"))
+		content = process(content, indentByFn(indent))
+		_, err := s.writer.Write([]byte(color.Faint.Sprint(content + "\n")))
```
Comment (efekarakus) [inline internal/pkg/template/diff/write.go:47]: nicesu

File: internal/pkg/template/diff/diff.go [hunk 2]
```diff
@@ -159,11 +184,19 @@ func parseSequence(fromNode, toNode *yaml.Node) ([]diffNode, error) {
 		return nil, nil
 	}
 	var children []diffNode
+	var matchCount int
 	inspector := newLCSStateMachine(fromSeq, toSeq, lcsIndices)
 	for action := inspector.action(); action != actonDone; action = inspector.action() {
```
Comment (efekarakus) [inline internal/pkg/template/diff/diff.go]: :lolsob: there is a little typo here 
```suggestion
	for action := inspector.action(); action != actionDone; action = inspector.action() {
```

File: internal/pkg/template/diff/diff.go [hunk 3]
```diff
@@ -159,11 +184,19 @@ func parseSequence(fromNode, toNode *yaml.Node) ([]diffNode, error) {
...
+			continue
+		}
+		if matchCount != 0 {
+			children = append(children, &unchangedNode{count: matchCount})
+			matchCount = 0
+		}
 		switch action {
-		case actionMatch:
-			// TODO(lou1415926): (x unchanged items)
 		case actionMod:
```
Comment (efekarakus) [inline internal/pkg/template/diff/diff.go]: I wonder if we could keep the nice old structure, if we have a flow like this:
```go
case actionMatch:
    matchCount += 1
    if action := inspector.peek(); action != actionMatch { // `peek` allows you to see the next action, without calling next()
         children = append(children, &unchangedNode{count: matchCount})
         matchCount = 0
    }
```
I think that should get rid of the last if-statement too, what do you think?

File: internal/pkg/template/diff/diff.go [hunk 4]
```diff
@@ -159,11 +184,19 @@ func parseSequence(fromNode, toNode *yaml.Node) ([]diffNode, error) {
 		return nil, nil
 	}
 	var children []diffNode
+	var matchCount int
 	inspector := newLCSStateMachine(fromSeq, toSeq, lcsIndices)
 	for action := inspector.action(); action != actonDone; action = inspector.action() {
+		if action == actionMatch {
+			matchCount++
+			inspector.next()
```
Comment (huanjani) [inline internal/pkg/template/diff/diff.go]: cool. I didn't know about the inspector package previously.

Comment (Lou1415926) [reply internal/pkg/template/diff/diff.go]: that is indeed super new! from https://github.com/aws/copilot-cli/pull/4611/files#diff-f9820ec569ed22a718e8dcacc806ef07b1455aca9562c18d975a5315b03d8400

File: internal/pkg/template/diff/diff.go (hunk 3 above)
Comment (Lou1415926) [reply internal/pkg/template/diff/diff.go]: i think this is a very good idea boss

File: internal/pkg/template/diff/iter.go [hunk 5]
```diff
@@ -76,6 +78,14 @@ func (i *lcsStateMachine) action() action {
 	return action
 }
 
+func (i *lcsStateMachine) peek() action {
```
Comment (efekarakus) [inline internal/pkg/template/diff/iter.go]: nit: (for all receivers)
```suggestion
func (sm *lcsStateMachine) peek() action {
```

**Review by efekarakus — APPROVED** [review]
🚀

File: internal/pkg/template/diff/iter.go [hunk 6]
```diff
@@ -56,53 +58,61 @@ func (i *lcsStateMachine) action() action {
...
 	}
-	i.currAction = action
+	sm.currAction = action
 	return action
 }
 
-func (i *lcsStateMachine) next() {
-	switch i.currAction {
+func (sm *lcsStateMachine) peek() action {
+	var lcsIdxOld, fromIdxOld, toIdxOld, actionOld = sm.lcsIndices.index, sm.from.index, sm.to.index, sm.currAction
```
Comment (dannyrandall) [inline internal/pkg/template/diff/iter.go]:
```suggestion
	lcsIdxOld, fromIdxOld, toIdxOld, actionOld := sm.lcsIndices.index, sm.from.index, sm.to.index, sm.currAction
```
nit

File: internal/pkg/template/diff/write.go [hunk 7]
```diff
@@ -38,9 +40,12 @@ func (s *treeWriter) write() error {
 
 func (s *treeWriter) writeTree(node diffNode, indent int) error {
 	var formatter formatter
-	switch node.(type) {
-	// case *unchangedNode:
-	// TODO(lou1425926): handle unchanged. 
+	switch node := node.(type) {
+	case *unchangedNode:
+		content := fmt.Sprintf("(%s)", english.Plural(node.unchangedCount(), "unchanged item", "unchanged items"))
```
Comment (dannyrandall) [inline internal/pkg/template/diff/write.go:45]:
```suggestion
		content := fmt.Sprintf("(%d)", english.Plural(node.unchangedCount(), "unchanged item", "unchanged items"))
```

**Review by dannyrandall — APPROVED** [review]
lgtm!

File: internal/pkg/template/diff/write.go (hunk 7 above)
Comment (Lou1415926) [reply internal/pkg/template/diff/write.go:45]: `english.Plural` returns a string that looks like `"1 unchanged item"` or `"x unchanged items"`!

Comment (dannyrandall) [reply internal/pkg/template/diff/write.go:45]: oooh i see now! cool 👍

**Pull request merged**